RFC.md

```diff
@@ -0,0 +1,466 @@
+        fusion_layer (nn.Module): new fusion layer
+        fusion_first (bool): boolean to insert fusion layer before or after
+	        the decoder layer.
+    """
+
+    def __init__(
+        self,
+        layer: nn.Module,
+        fusion_layer: nn.Module,
+        fusion_first: bool = True
+    ):
+        super().__init__()
+        self.layer = layer
+        self.fusion_layer = fusion_layer
+        self.fusion_first = fusion_first
+        self.register_state_dict_post_hook(Test._state_dict_hook)
+        # TODO: Switch to register_load_state_dict_pre_hook in v2.5
+        self._register_load_state_dict_pre_hook(
+	        Test._load_state_dict_hook,
+	        with_module=True
+	    )
+
+    def _state_dict_hook(self, destination, prefix, keep_vars):
+	    """
+		Remove "layer" from the original layer in the state_dict
+		name. This keeps the orginal state dict name for the layer
+		from before fusing with the second layer.
+
+		[!Note] This update changes the order of the OrderedDict
+		"""
+        keys = list(destination.keys())
+        for key in keys:
+            if key.startswith("layer"):
+                new_key = key.replace("layer.", "")
+                destination[new_key] = destination[key]
+                del destination[key]
+
+    def _load_state_dict_hook(self, state_dict, *args, **kwargs):
+	    """ Undo the change from _state_dict_hook """
+        keys = list(state_dict.keys())
+        for key in keys:
+            if not key.startswith("fusion_layer"):
+                new_key = "layer." + key
+                state_dict[new_key] = state_dict[key]
+                del state_dict[key]
+
+	def setup_cache(self, batch_size: int, dtype: torch.dtype) -> None:
+		self.layer.setup_cache(batch_size, dtype)
+		self.fusion_layer.setup_cache(batch_size, dtype)
+
+    def cache_enabled(self) -> bool:
+        return self.layer.cache_enabled()
+
+    def reset_cache(self):
+        self.layer.reset_cache()
+        self.fusion_layer.reset_cache()
+
+    def fusion_params(self) -> List[str]:
+        """
+        Return parameters of fusion layer.
+        """
+        fusion_params = [
+            f"fusion_layer.{k}" for k, v in self.fusion_layer.named_parameters()
+        ]
+        return fusion_params
+
+    def forward(self, x: Tensor, **kwargs) -> Tensor:
+        """
+        Args:
+            x (Tensor): input tensor with shape
+                [batch_size x seq_length x embed_dim]
+
+        Returns:
+            Tensor: output tensor with same shape as input
+                [batch_size x seq_length x embed_dim]`
+
+        """
+        if self.fusion_first:
+            x = self.fusion_layer(x, **kwargs)
+            x = self.layer(x, **kwargs)
+        else:
+            x = self.layer(x, **kwargs)
+            x = self.fusion_layer(x, **kwargs)
+        return x
+```
+
+### Fusion Embedding
+The reason for this layer is summarized in the docstring.
+
+```python
+class FusionEmbedding(nn.Module):
```

Comment:
This is very interesting, can honestly probably extend this beyond just multimodal to any new special tokens, right? (Looking at you Llama3 instruct..) If that's the case we should think about how to support different initialization schemes for the new tokens too

Reply:
Yeah this isn't specific to multimodal, it's just an efficient way to extend number of input tokens. Though it only works if you never want the model to be able to output those tokens.